content/courses/intro-to-embedded-sys-prog/chapters/handling_interrupts.rst

```diff
@@ -6,177 +6,1537 @@ Handling Interrupts
+   begin
+      Next_Release := Clock;
+      loop
+         exit when Status (...);
+
+         Next_Release := Next_Release + Period;
+         delay until Next_Release;
+      end loop;
+   end Poll_With_Delay;
+
+The code above will check the status of some device every 30
+milliseconds (an arbitrary period just for illustration) until the
+:ada:`Status` function result allows the loop to exit. If the device
+"hangs" the loop is never exited, but as you saw there are ways to
+address that possibility. When the code does not exit the loop, the next
+point on the timeline is computed and the task executing the code then
+suspends, allowing the other tasks in the application to execute.
+Eventually, the next release point is reached and so the task becomes
+ready to execute again (and will, subject to priorities).
+
+But how long should the polling task suspend when awaiting the device?
+We need to suspend long enough for the other tasks to get something
+done, but not so long that the device isn't handled fast enough. Finding
+the right balance is often not simple, and is further complicated by the
+"task switching" time. That's the time it takes to switch the execution
+context from one task to another, in this case in response to the "delay
+until" statement suspending the polling task. And it must be considered
+in both directions: when the delay expires we'll eventually switch back
+to the polling task.
+
+As you can see, polling is easily expressed but has potentially
+significant drawbacks and architectural ramifications so it should be
+avoided as a default approach.
+
+Now let's explore the Ada interrupt facilities.
+
+
+Language-Defined Interrupt Model
+--------------------------------
+
+The Ada language standard defines a model for hardware interrupts, as
+well as language-defined mechanisms for handling interrupts consistent
+with that model. The model is defined in Annex C, the "Systems
+Programming" annex, section 3 "Interrupt Support." The following is the
+text of that section with only a few simplifications and elisions.
+
+* Interrupts are said to occur. An occurrence of an interrupt is separable
+  into generation and delivery.
+
+  - Generation of an interrupt is the event in
+    the underlying hardware or system that makes the interrupt available to
+    the program.
+
+  - Delivery is the action that invokes part of the program as
+    response to the interrupt occurrence.
+
+* Between generation and delivery, the interrupt occurrence is pending.
+
+* Some or all interrupts may be blocked. When an interrupt is blocked, all
+  occurrences of that interrupt are prevented from being delivered.
+
+* Certain interrupts are reserved. A reserved interrupt is either an
+  interrupt for which user-defined handlers are not supported, or one
+  which already has an attached handler by some other
+  implementation-defined means. The set of reserved interrupts is
```

Comment:
As discussed separately, "implementation-defined" could be confusing. Instead, whenever appropriate, we should make it clear that we're referring to the compiler implementation and use an expression such as "defined in the compiler implementation," "compiler-specific," or "implemented in the compiler."

Reply:
Agreed, and updated.